### PR TITLE
Migrate node-opendds from Travis CI to Github Actions

### PR DESCRIPTION
**Pull request by lij-oci**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      - run: export V8_ROOT=$npm_config_devdir/$(node -v | cut -c 2-)
+      - run: export NAN_ROOT=$BUILD_DIR/node_modules/nan
+      - run: $ACE_ROOT/bin/mwc.pl -type gnuace -exclude ACE_TAO,OpenDDS
+      - run: export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$BUILD_DIR/tools/v8stubs
+      - run: make
+      - run: if [ "$DDS_SECURITY" == 1 ]; then TEST_OPTIONS+="--secure"; fi
+      - run: cd test;
+      - run: ./run_test.pl cpp2node $TEST_OPTIONS
+      - run: ./run_test.pl node2cpp $TEST_OPTIONS
+      - run: ./run_test.pl node2node $TEST_OPTIONS
```
Comment (iguessthislldo) [inline .github/workflows/build.yml:66]: Replace this with the multi line YAML syntax. I'm not even sure this is going to work correctly as is.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        name: Install dependencies (Ubuntu)
+        run: |-
+          sudo apt-get update
+          sudo apt-get -y install libssl-dev libxerces-c-dev
+      - name: 'Set up Node.js ${{ matrix.node-version }}'
+        uses: actions/setup-node@v1
+        with:
+          node-version: ${{ matrix.node-version }}
+      - uses: actions/checkout@v2
+        run: |
```
Comment (iguessthislldo) [inline .github/workflows/build.yml:40]:
```suggestion
        run: |-
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -51,19 +43,28 @@ jobs:
...
-          fi
-          echo "dds_security=${{ matrix.dds_security }}; CONFIG_OPTIONS=$CONFIG_OPTIONS; BUILD_TARGETS=$BUILD_TARGETS"
+          if [ ${{ matrix.dds_security }} == 1 ]; then CONFIG_OPTIONS+="--security"; fi
+          echo "dds_security=${{ matrix.dds_security }}; CONFIG_OPTIONS=$CONFIG_OPTIONS"
           cd OpenDDS
           ./configure --no-tests --std=c++11 $CONFIG_OPTIONS
           cd ..
-          make -C OpenDDS -sj6 DCPSInfoRepo_Main OpenDDS_Rtps_Udp $BUILD_TARGETS
+      - run: |-
+          export npm_config_devdir=$GITHUB_WORKSPACE/opendds-node-gyp-devdir
```
Comment (iguessthislldo) [inline .github/workflows/build.yml]:
```suggestion
          export npm_config_devdir=$GITHUB_WORKSPACE/opendds-node-gyp-devdir
          echo "npm_config_devdir=$npm_config_devdir" >> $GITHUB_ENV
```

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+          echo "NAN_ROOT=$GITHUB_WORKSPACE/node_modules/nan" >> $GITHUB_ENV
+          echo "npm_config_devdir=$GITHUB_WORKSPACE/opendds-node-gyp-devdir" >> $GITHUB_ENV
+          echo "LD_LIBRARY_PATH=$GITHUB_WORKSPACE/ACE_TAO/ACE/lib:$GITHUB_WORKSPACE/OpenDDS/lib:$GITHUB_WORKSPACE/tools/v8stubs:$LD_LIBRARY_PATH" >> $GITHUB_ENV
+      - name: 'Install node'
+        run: |-
+          git clone https://github.com/nvm-sh/nvm.git
+          cd ${{ env.NVM_DIR }}
+          git checkout v0.38.0
+          [ -s "./nvm.sh" ] && \. "./nvm.sh"  # loads nvm
+          [ -s "./bash_completion" ] && \. "./bash_completion"  # loads nvm bash_completion
```
Comment (iguessthislldo) [inline .github/workflows/build.yml:44]:
```suggestion
```

There is no need for bash completion in a script like this.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+          echo "TAO_ROOT=$GITHUB_WORKSPACE/ACE_TAO/TAO" >> $GITHUB_ENV
+          echo "DDS_ROOT=$GITHUB_WORKSPACE/OpenDDS" >> $GITHUB_ENV
+          echo "MPC_ROOT=$GITHUB_WORKSPACE/MPC" >> $GITHUB_ENV
+          echo "NVM_DIR=$GITHUB_WORKSPACE/nvm" >> $GITHUB_ENV
+          echo "NAN_ROOT=$GITHUB_WORKSPACE/node_modules/nan" >> $GITHUB_ENV
+          echo "npm_config_devdir=$GITHUB_WORKSPACE/opendds-node-gyp-devdir" >> $GITHUB_ENV
+          echo "LD_LIBRARY_PATH=$GITHUB_WORKSPACE/ACE_TAO/ACE/lib:$GITHUB_WORKSPACE/OpenDDS/lib:$GITHUB_WORKSPACE/tools/v8stubs:$LD_LIBRARY_PATH" >> $GITHUB_ENV
+      - name: 'Clone OpenDDS'
+        run: |-
+          git clone --depth 1 git://github.com/DOCGroup/MPC.git
```
Comment (simpsont-oci) [inline .github/workflows/build.yml:42]: Let's break each of these clones into separate checkout steps and use the checkout@v2 action.

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+          git clone --depth 1 -b Latest_Micro git://github.com/DOCGroup/ACE_TAO.git
+          git clone --depth 1 -b ${{ matrix.opendds_branch }} git://github.com/objectcomputing/OpenDDS.git
+          if [ ${{ matrix.m.dds_security }} == 1 ]; then
+            echo 'Install ssl and xerces'
+            sudo apt-get update
```
Comment (simpsont-oci) [inline .github/workflows/build.yml:47]: move these installs to a separate step... you can make the step conditional (similar to this if statement).

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+          [ -s "./nvm.sh" ] && \. "./nvm.sh"  # loads nvm
+          cd $GITHUB_WORKSPACE
+          nvm install ${{ matrix.node_version }}
+          nvm use ${{ matrix.node_version }}
+          npm install --save nan
+          npm install -g node-gyp
+          node-gyp configure build
+          echo "node:$(node -v)  npm:$(npm -v)"
+          export V8_ROOT=${{ env.npm_config_devdir }}/$(node -v | cut -c 2-)
+          echo "Build ========== ========== =========="
```
Comment (simpsont-oci) [inline .github/workflows/build.yml:75]: Let's break the build (make) out into a separate step and add the problem matcher (see OpenDDS build_and_test.yml) to catch compile warnings for building node-opendds.

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+          ${{ env.ACE_ROOT }}/bin/mwc.pl -type gnuace -exclude ACE_TAO,OpenDDS
+          make
+          echo "Test ========== ========== =========="
```
Comment (simpsont-oci) [inline .github/workflows/build.yml:78]: And let's have test be its own step as well.

**Review by simpsont-oci — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 7 above)
Comment (lij-oci) [reply .github/workflows/build.yml:90]: I tried to separate these 3 steps: 'Install node', 'Build', and 'Test' from early on, see commit 'separate steps 18cd1bb' 14 days ago. It didn't work, until 4 days ago, I created this step: 'Install node, Build, Test' (commits: 781a4c7, 8acbaf3, a0835f8). When it worked, I tried to separate the 'Test' step (commit: 791bbdb), and it failed again. I don't like these 3 steps in one, but until we solve the node version with OpenDDS branch/version issues, the separate steps here are not likely to work.

File: .github/workflows/build.yml (hunk 8 above)
Comment (lij-oci) [reply .github/workflows/build.yml:93]: see above.

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        with:
+          repository: objectcomputing/OpenDDS
+          ref: ${{ matrix.opendds_branch }}
+          path: OpenDDS
+          fetch-depth: 1
+      - name: 'Install ssl and xerces'
+        run: |-
+          sudo apt-get update
+          sudo apt-get -y install libssl-dev libxerces-c-dev
+        if: ${{ matrix.m.dds_security == 1}}
```
Comment (simpsont-oci) [inline .github/workflows/build.yml:64]:
```suggestion
        if: ${{ matrix.m.dds_security == 1 }}
```

**Review by simpsont-oci — APPROVED** [review]

**Pull request merged**